sdk/src/main/scala/com/horizen/account/state/AccountStateView.scala

```diff
@@ -179,6 +179,16 @@ class AccountStateView(metadataStorageView: AccountStateMetadataStorageView,
     BigInteger.ZERO
   }
 
+  def applyMessage(message: Message): Option[ExecutionResult] = {
+    // TODO: some refactoring is required here, applyMessage should include:
+    //  - buying gas
+    //  - validations that the caller has enough funds for everything (gas and value transfer)
+    //  - processing message
+    //  - returning unused gas and refunds
```

Comment:
This point requires a lot of attention.
Consider that `eth_estimateGas` uses staticCalll to predict the gasLimit, it expects the result before `refund`.
Otherwise, as for me, the gasLimit is going to be too small, so the real Tx will fail.

Reply:
Interesting point. I had a look at the geth implementation and they do not return the gas used before refund.

They do a binary search to find a gasLimit that allows the transaction to succeed (with starting bounds `low=21000` and `high=args.gasLimit`).

Wallets will also not use the result of `eth_estimateGas` directly as the `gasLimit`, but rather add a buffer on top. In the source of MetaMask I found a [multiplier of `1.5`](https://github.com/MetaMask/metamask-extension/blob/c72199a1a6e4151c40c22f79d0f3b6ed7a2d59a7/app/scripts/controllers/transactions/tx-gas-utils.js#L86).

Reply:
Strange but ok.
Let's consider it in the current tasks